File: Zeze/Util/SimpleThreadPool.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Collections.Concurrent;

namespace Zeze.Util
{
    public sealed class SimpleThreadPool
    {
        private static readonly NLog.Logger logger = NLog.LogManager.GetCurrentClassLogger();

        private readonly BlockingCollection<Action> taskQueue = new BlockingCollection<Action>();
        private readonly List<Thread> workers = new List<Thread>();
        public string Name { get; }

        public int AliveThreadCount
        {
            get
            {
                int count = 0;
                foreach (var worker in workers)
                {
                    if (worker.IsAlive)
                        count++;
                }
                return count;
            }
        }

        public bool QueueUserWorkItem(Action action)
        {
            return taskQueue.TryAdd(action);
        }

        public void Shutdown(bool immediately = true)
        {
            lock (this)
            {
                if (immediately)
                {
                    foreach (var worker in workers)
                        taskQueue.Add(null);
                    taskQueue.CompleteAdding();
                }
                else
                {
                    taskQueue.CompleteAdding();
                    while (taskQueue.IsCompleted == false)
                    {
                        Monitor.Wait(this);
                    }
                }
            }
        }

        public SimpleThreadPool(int workerThreads, string poolName)
        {
            Name = poolName;

            for (int i = 0; i < workerThreads; ++i)
            {
                workers.Add(new Thread(MainRun)
                {
                    Name = $"{Name}.{i}",
                    IsBackground = true
                });
            }

            foreach (Thread thread in workers)
            {
                thread.Start();
            }

        }

        private void MainRun()
        {
            while (true)
            {
                Action action = null;
                try
                {
                    lock (this)
                    {
                        if (taskQueue.IsCompleted)
                        {
                            Monitor.PulseAll(this);
                            break;
                        }
                    }
                    action = taskQueue.Take();
                    if (null == action)
                        break;
                    action();
                }
                /*
                catch (OperationCanceledException)
                {
                    // skip
                }
                */
                catch (Exception ex)
                {
                    logger.Error(ex, "SimpleThreadPool {0}", action);
                }
            }
        }
    }
}

File: ZezeJava/src/Zeze/Util/SimpleThreadPool.java
package Zeze.Util;

import Zeze.*;
import java.util.*;

public final class SimpleThreadPool {
	private static final NLog.Logger logger = NLog.LogManager.GetCurrentClassLogger();

	private BlockingCollection<tangible.Action0Param> taskQueue = new BlockingCollection<tangible.Action0Param>();
	private ArrayList<Thread> workers = new ArrayList<Thread>();
	private String Name;
	public String getName() {
		return Name;
	}

	public boolean QueueUserWorkItem(tangible.Action0Param action) {
		if (false == taskQueue.IsAddingCompleted) {
			taskQueue.Add(action);
		}
		// skip task
		return true;
	}

	public void Shutdown() {
		taskQueue.CompleteAdding();

		synchronized (this) {
			while (taskQueue.IsCompleted == false) {
				Monitor.Wait(this);
			}
		}
	}

	public SimpleThreadPool(int workerThreads, String poolName) {
		Name = poolName;

		for (int i = 0; i < workerThreads; ++i) {
			workers.add(new Thread() {
			void run() {
				MainRun();
			}
			};
		}

		for (Thread thread : workers) {
			thread.start();
		}

	}

	private void MainRun() {
		while (true) {
			tangible.Action0Param action = ::null;
			try {
				if (taskQueue.IsCompleted) {
					synchronized (this) {
						Monitor.PulseAll(this);
						break;
					}
				}
				action = taskQueue.Take();
				if (null == action) {
					break;
				}
				action.invoke();
			}
			/*
			catch (OperationCanceledException)
			{
			    // skip
			}
			*/
			catch (RuntimeException ex) {
				logger.Error(ex, "SimpleThreadPool {0}", action);
			}
		}
	}
}